recipes-extended/xen/blktap3/fix-run-time-errors-and-memory-leaks.patch

```diff
@@ -0,0 +1,270 @@
+################################################################################
+SHORT DESCRIPTION: 
+################################################################################
+Resolve run-time errors and memory leaks.
+
+################################################################################
+LONG DESCRIPTION: 
+################################################################################
+Resolve run-time errors and memory leaks.
+
+################################################################################
+CHANGELOG 
+################################################################################
+Author(s):
+Mahantesh Salimath <salimathm@ainfosec.com>
+
+################################################################################
+REMOVAL 
+################################################################################
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+Yes, should be upstreamed.
+
+################################################################################
+INTERNAL DEPENDENCIES 
+################################################################################
+
+################################################################################
+PATCHES 
+################################################################################
+Index: blktap/control/tap-ctl-free.c
+===================================================================
+--- blktap.orig/control/tap-ctl-free.c
++++ blktap/control/tap-ctl-free.c
+@@ -33,6 +33,7 @@
+ #endif
+ 
+ #include <stdio.h>
++#include <string.h>
+ #include <errno.h>
+ #include <fcntl.h>
+ #include <stdlib.h>
+@@ -48,6 +49,20 @@ tap_ctl_free(const int minor)
+ {
+ 	int fd, err;
+ 
++	char *tapdev_name, *ring_name;
++	err = asprintf(&tapdev_name, "%s%d",
++		       BLKTAP2_IO_DEVICE, minor);
++	if(err == -1){
++		err = ENOMEM;
++		return err;
++	}
++	err = asprintf(&ring_name, "%s%d",
++		       BLKTAP2_RING_DEVICE, minor);
++	if(err == -1){
++		err = ENOMEM;
++		return err;
++	}
++
+ 	fd = open(BLKTAP2_CONTROL_DEVICE, O_RDONLY);
+ 	if (fd == -1) {
+ 		EPRINTF("failed to open control device: %d\n", errno);
+@@ -57,5 +72,10 @@ tap_ctl_free(const int minor)
+ 	err = ioctl(fd, BLKTAP2_IOCTL_FREE_TAP, minor);
+ 	close(fd);
+ 
++	unlink(tapdev_name);
++	unlink(ring_name);
++	free(tapdev_name);
++	free(ring_name);
++
+ 	return err ? -errno : 0;
+ }
+Index: blktap/control/tap-ctl-ipc.c
+===================================================================
+--- blktap.orig/control/tap-ctl-ipc.c
++++ blktap/control/tap-ctl-ipc.c
+@@ -64,7 +64,7 @@ tap_ctl_read_raw(int fd, void *buf, size
+ 		if (ret == -1)
+ 			break;
+ 		else if (FD_ISSET(fd, &readfds)) {
+-			ret = read(fd, buf + offset, size - offset);
++			ret = read(fd, (char*)buf + offset, size - offset);
```

Comment:
I thought you could add an offset to a void * pointer without an issue.  i.e. is this cast needed?

Reply:
I'm pretty sure I got a segfault for not using the cast.